common/src/main/java/io/netty/util/ResourceLeakDetector.java

```diff
@@ -428,121 +450,138 @@ public boolean close(T trackedObject) {
 
-        private static final Set<String> STACK_TRACE_ELEMENT_EXCLUSIONS = new HashSet<String>();
-        static {
-            STACK_TRACE_ELEMENT_EXCLUSIONS.add("io.netty.util.ReferenceCountUtil.touch");
-            STACK_TRACE_ELEMENT_EXCLUSIONS.add("io.netty.buffer.AdvancedLeakAwareByteBuf.touch");
-            STACK_TRACE_ELEMENT_EXCLUSIONS.add("io.netty.buffer.AbstractByteBufAllocator.toLeakAwareBuffer");
-            STACK_TRACE_ELEMENT_EXCLUSIONS.add(
-                    "io.netty.buffer.AdvancedLeakAwareByteBuf.recordLeakNonRefCountingOperation");
+    public static void addExclusions(Class clz, String ... methodNames) {
+        Set<String> nameSet = new HashSet<String>(Arrays.asList(methodNames));
+        // Use loop rather than lookup.  This avoids knowing the parameters, and doesn't have to handle
+        // NoSuchMethodException.
+        for (Method method : clz.getDeclaredMethods()) {
+            if (nameSet.remove(method.getName())) {
+                if (nameSet.isEmpty()) {
+                    break;
+                }
+            }
         }
+        if (!nameSet.isEmpty()) {
+            throw new IllegalArgumentException("Can't find '" + nameSet + "' in " + clz.getName());
+        }
+        String[] oldMethods;
+        String[] newMethods;
+        do {
+            oldMethods = excludedMethods.get();
+            newMethods = Arrays.copyOf(oldMethods, oldMethods.length + 2 * methodNames.length);
+            for (int i = 0; i < methodNames.length; i++) {
+                newMethods[oldMethods.length + i * 2] = clz.getName();
+                newMethods[oldMethods.length + i * 2 + 1] = methodNames[i];
+            }
+        } while (!excludedMethods.compareAndSet(oldMethods, newMethods));
+    }
+
+    private static final class Record extends Throwable {
+
+        private static final Record BOTTOM = new Record();
 
         private final String hintString;
-        Record next;
+        private final Record next;
+        private final int pos;
 
-        Record(Object hint) {
+        Record(Record next, Object hint) {
             // This needs to be generated even if toString() is never called as it may change later on.
             hintString = hint instanceof ResourceLeakHint ? ((ResourceLeakHint) hint).toHintString() : hint.toString();
+            this.next = next;
+            this.pos = next.pos + 1;
         }
 
-        Record() {
+        Record(Record next) {
            hintString = null;
+           this.next = next;
+           this.pos = next.pos + 1;
+        }
+
+        // Used to terminate the stack
+        private Record() {
+            hintString = null;
+            next = null;
+            pos = -1;
         }
 
         @Override
         public String toString() {
-            StringBuilder buf = new StringBuilder(4096);
+            StringBuffer buf = new StringBuffer(2048);
```

Comment:
![MAJOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-major.png) Replace the synchronized class "StringBuffer" by an unsynchronized one such as "StringBuilder". [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](https://garage.netty.io/sonarqube/coding_rules#rule_key=squid%3AS1149)

Reply:
@carl-mastrangelo I think @netkins is right here... you want StringBuilder

Reply:
yep, I was thrown by the name being "buf"   In a previous iteration a single buffer was passed around, but SB is not powerful enough to record indices to check for duplicates.  It ended up back as String.